bugsnag-android-core/src/main/java/com/bugsnag/android/ReportDeliveryDelegate.java

```diff
@@ -0,0 +1,109 @@
+package com.bugsnag.android;
+
+import androidx.annotation.NonNull;
+
+import java.util.Collections;
+import java.util.Date;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.RejectedExecutionException;
+
+class ReportDeliveryDelegate extends BaseObservable {
+
+    final Logger logger;
+    final EventStore eventStore;
+    final ImmutableConfig immutableConfig;
+    final BreadcrumbState breadcrumbState;
+
+    ReportDeliveryDelegate(Logger logger, EventStore eventStore,
+                           ImmutableConfig immutableConfig, BreadcrumbState breadcrumbState) {
+        this.logger = logger;
+        this.eventStore = eventStore;
+        this.immutableConfig = immutableConfig;
+        this.breadcrumbState = breadcrumbState;
+    }
+
+    void deliverEvent(@NonNull Event event) {
+        // Build the report
+        Report report = new Report(immutableConfig.getApiKey(), event);
+        Session session = event.getSession();
+
+        if (session != null) {
+            setChanged();
+
+            if (event.isUnhandled()) {
+                event.setSession(session.incrementUnhandledAndCopy());
+                notifyObservers(StateEvent.NotifyUnhandled.INSTANCE);
+            } else {
+                event.setSession(session.incrementHandledAndCopy());
+                notifyObservers(StateEvent.NotifyHandled.INSTANCE);
+            }
+        }
+
+        if (event.isUnhandled()) {
+            eventStore.write(event);
+            eventStore.flushAsync();
+        } else {
+            deliverReportAsync(event, report);
+        }
+    }
+
+    private void deliverReportAsync(@NonNull Event event, Report report) {
+        final Report finalReport = report;
+        final Event finalEvent = event;
+
+        // Attempt to send the report in the background
+        try {
+            Async.run(new Runnable() {
+                @Override
+                public void run() {
+                    deliver(finalReport, finalEvent);
+                }
+            });
+        } catch (RejectedExecutionException exception) {
+            eventStore.write(event);
+            logger.w("Exceeded max queue count, saving to disk to send later");
+        }
+    }
+
+    DeliveryStatus deliver(@NonNull Report report, @NonNull Event event) {
+        DeliveryParams deliveryParams = immutableConfig.errorApiDeliveryParams();
```

Comment:
Not a change from this PR directly, but should these functions be named `getErrorApiDeliveryParams` and `getSesssionApiDeliveryParams`?